### PR TITLE
platform: add OpenStack

### PR DESCRIPTION
**Pull request by arithx**
Adds the OpenStack platform.

This currently requires a patch (coreos/coreos-overlay#3415) for `coreos.ignition.v2_2.security.tls` to pass.

### CODE REVIEW

**Comment by arithx** [comment]
Could not rebase this cleanly as even on just current master a `make vendor` results in changes that @ajeddeloh was going to investigate sometime soon.

File: README.md [hunk 1]
```diff
@@ -246,6 +246,25 @@ you can paste in. This will populate the `.boto` file.
 See [Google Cloud Platform's Documentation](https://cloud.google.com/storage/docs/boto-gsutil)
 for more information about the `.boto` file.
 
+### openstack
+`openstack` uses `~/.config/openstack.json`. This can be configured manually:
```
Comment (cgwalters) [inline README.md]: Do any other tools parse this or is it a kola invention?

File: cmd/ore/openstack/create.go [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+	}
+
+	path string
+	name string
+)
+
+func init() {
+	OpenStack.AddCommand(cmdCreate)
+	cmdCreate.Flags().StringVar(&path, "file",
+		sdk.BuildRoot()+"/images/amd64-usr/latest/coreos_production_openstack_image.img",
```
Comment (cgwalters) [inline cmd/ore/openstack/create.go]: Just a note, we should start thinking about how this type of stuff intersects with coreos-assembler which has a different build layout and naming convention.  But doesn't block this.

**Comment by ashcrow** [comment]
*bump*

File: auth/openstack.go [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+	"fmt"
+	"os"
+	"os/user"
+	"path/filepath"
+)
+
+const OpenStackConfigPath = ".config/openstack.json"
+
+type OpenStackProfile struct {
+	AuthURL    string `json:"os_auth_url"`
```
[the remark this answers is not in the document]
Comment (bgilbert) [reply auth/openstack.go]: Those are uppercase and... not wrapped in JSON.  So I don't think it's especially valuable to keep the prefixes.  I'm okay with it if you feel strongly, though.

[the remark this answers is not in the document]
Comment (arithx) [reply auth/openstack.go]: I don't feel particularly strongly either way, I'll drop the prefix.

File: platform/api/openstack/api.go [hunk 4]
```diff
@@ -0,0 +1,588 @@
...
+	// Region (e.g. "regionOne")
+	Region string
+	// Instance Flavor ID
+	Flavor string
+	// Image ID
+	Image string
+	// Network ID
+	Network string
+	// Domain ID
+	DomainID string
```
[the remark this answers is not in the document]
Comment (bgilbert) [reply platform/api/openstack/api.go]: `OpenStackProfile.DomainID` is passed into `AuthOptions`.  `Options.DomainID` is unused.

[the remark this answers is not in the document]
Comment (arithx) [reply platform/api/openstack/api.go]: Ah yeah I see that now, I need to add a section to allow overriding it via the command line option.

**Comment by arithx** [comment]
Updated based on comments, rebased, and pushed :arrow_up:

File: cmd/ore/openstack/openstack.go [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+
+	OpenStack = &cobra.Command{
+		Use:   "openstack [command]",
+		Short: "OpenStack machine utilities",
+	}
+
+	API     *openstack.API
+	options openstack.Options
+
+	imageName string
```
Comment (bgilbert) [inline cmd/ore/openstack/openstack.go]: Unused.

File: README.md [hunk 6]
```diff
@@ -206,7 +206,7 @@ kola spawn -p aws --aws-profile other_profile
 ```
 
 ### aws
-`aws` reads the `~/.aws/credentials` file used by Amazon's aws command-line tool.
+`aws` reads the `~/.aws/credentials` file used by Amazon's aws command line tool.
```
Comment (bgilbert) [inline README.md]: It's hyphenated when used as an adjective.  :slightly_smiling_face:

File: README.md [hunk 7]
```diff
@@ -261,6 +261,25 @@ you can paste in. This will populate the `.boto` file.
 See [Google Cloud Platform's Documentation](https://cloud.google.com/storage/docs/boto-gsutil)
 for more information about the `.boto` file.
 
+### openstack
+`openstack` uses `~/.config/openstack.json`. This can be configured manually:
+```
+{
+    "default": {
+        "os_auth_url": "auth url here",
```
Comment (bgilbert) [inline README.md]: `os_` prefixes are gone now.

File: README.md [hunk 8]
```diff
@@ -261,6 +261,25 @@ you can paste in. This will populate the `.boto` file.
...
+        "os_tenant_id": "tenant id here",
+        "os_tenant_name": "tenant name here",
+        "os_username": "username here",
+        "os_password": "password here",
+        "os_user_domain_id": "domain id here",
```
Comment (bgilbert) [inline README.md]: `user_domain`

File: README.md [hunk 9]
```diff
@@ -261,6 +261,25 @@ you can paste in. This will populate the `.boto` file.
...
+        "os_floating_ip_pool": "floating ip pool here",
+        "os_region_name": "region here"
+    }
+}
+```
+
+`os_user_domain_id` is required on some newer versions of OpenStack using Keystone V3 but is optional on older versions. `os_floating_ip_pool` and `os_region_name` can be optionally specified here to be used as a default if not specified on the command line.
```
Comment (bgilbert) [inline README.md]: `user_domain`

File: platform/api/openstack/api.go [hunk 10]
```diff
@@ -0,0 +1,600 @@
...
+func (a *API) DeleteKey(name string) error {
+	return keypairs.Delete(a.computeClient, name).ExtractErr()
+}
+
+func (a *API) listServersWithMetadata(metadata map[string]string) ([]servers.Server, error) {
+	pager := servers.List(a.computeClient, servers.ListOpts{})
+
+	pagesPtr, err := pagerCheck(pager)
+	if err != nil {
+		return nil, fmt.Errorf("servers: %v", err)
```
[the remark this answers is not in the document]
Comment (arithx) [reply platform/api/openstack/api.go]: Went the route of adding an `allowEmpty` field to `unwrapPages` which will skip the `IsEmpty` check and using it here and in `findFloatingIP`

**Comment by arithx** [comment]
Updated.

**Pull request merged**